File: breath_trainer.js
# breath_trainer.js
/**
 * 🌬️ Breath Trainer – 4-7-8 Relaxation Technique (Node.js Edition)
 * Features: guided breathing, visual progress bars, session stats, custom timing, sound alerts
 */

const fs = require('fs');
const path = require('path');
const os = require('os');
const readline = require('readline');

// ─── Colors ──────────────────────────────────────────────────────────────────

const colors = {
    reset: '\x1b[0m',
    bright: '\x1b[1m',
    dim: '\x1b[2m',
    red: '\x1b[31m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    magenta: '\x1b[35m',
    cyan: '\x1b[36m',
};

const c = (str, color) => `${color}${str}${colors.reset}`;

// ─── Sound ──────────────────────────────────────────────────────────────────

function beep() {
    // Use system bell
    process.stdout.write('\x07');
}

// ─── Config ──────────────────────────────────────────────────────────────────

const CONFIG = {
    dataDir: path.join(os.homedir(), '.breath_trainer'),
    dataFile: 'stats.json',
    defaultInhale: 4,
    defaultHold: 7,
    defaultExhale: 8,
};

// ─── Stats Manager ──────────────────────────────────────────────────────────

class StatsManager {
    constructor() {
        this.dataDir = CONFIG.dataDir;
        this.dataFile = path.join(this.dataDir, CONFIG.dataFile);
        this.totalSessions = 0;
        this.totalCycles = 0;
        this.totalTimeSeconds = 0;
        this.sessions = [];
        this.customTiming = { inhale: CONFIG.defaultInhale, hold: CONFIG.defaultHold, exhale: CONFIG.defaultExhale };
        if (!fs.existsSync(this.dataDir)) fs.mkdirSync(this.dataDir, { recursive: true });
        this._load();
    }

    _load() {
        if (fs.existsSync(this.dataFile)) {
            try {
                const data = JSON.parse(fs.readFileSync(this.dataFile, 'utf8'));
                this.totalSessions = data.totalSessions || 0;
                this.totalCycles = data.totalCycles || 0;
                this.totalTimeSeconds = data.totalTimeSeconds || 0;
                this.sessions = data.sessions || [];
                this.customTiming = data.customTiming || { inhale: CONFIG.defaultInhale, hold: CONFIG.defaultHold, exhale: CONFIG.defaultExhale };
            } catch (_) {}
        }
    }

    save() {
        fs.writeFileSync(this.dataFile, JSON.stringify({
            totalSessions: this.totalSessions,
            totalCycles: this.totalCycles,
            totalTimeSeconds: this.totalTimeSeconds,
            sessions: this.sessions,
            customTiming: this.customTiming
        }, null, 2));
    }

    recordSession(cycles, duration) {
        this.totalSessions++;
        this.totalCycles += cycles;
        this.totalTimeSeconds += Math.floor(duration);
        this.sessions.push({
            date: new Date().toISOString(),
            cycles,
            durationSeconds: Math.floor(duration)
        });
        if (this.sessions.length > 50) this.sessions = this.sessions.slice(-50);
        this.save();
    }

    setTiming(inhale, hold, exhale) {
        this.customTiming = { inhale, hold, exhale };
        this.save();
    }
}

// ─── Breath Trainer ──────────────────────────────────────────────────────

class BreathTrainer {
    constructor() {
        this.rl = readline.createInterface({ input: process.stdin, output: process.stdout });
        this.stats = new StatsManager();
        this.inhale = this.stats.customTiming.inhale;
        this.hold = this.stats.customTiming.hold;
        this.exhale = this.stats.customTiming.exhale;
        this.soundOn = true;
    }

    _sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    _showStage(stage, duration, color, emoji) {
        return new Promise(async (resolve) => {
            const steps = Math.floor(duration * 2);
            for (let i = 0; i <= steps; i++) {
                const pct = (i / steps * 100);
                const filled = Math.floor(i / steps * 30);
                const bar = '█'.repeat(filled) + '░'.repeat(30 - filled);
                process.stdout.write(`\r  ${emoji} ${c(stage, color)} ${c(bar, color)} ${Math.floor(pct)}%`);
                await this._sleep(500);
            }
            console.log();
            resolve();
        });
    }

    async _breathingCycle() {
        const start = Date.now();
        if (this.soundOn) beep();
        await this._showStage('Breathe In...', this.inhale, 'blue', '🌬️');
        if (this.soundOn) beep();
        await this._showStage('Hold...', this.hold, 'yellow', '⏸️');
        if (this.soundOn) beep();
        await this._showStage('Exhale...', this.exhale, 'green', '🌊');
        return (Date.now() - start) / 1000;
    }

    async runSession(cycles = 4) {
        console.log(c(`\n🌬️ 4-7-8 Breathing Session – ${cycles} cycles`, colors.bright + colors.cyan));
        if (this.soundOn) beep();
        console.log(c('Get ready...', colors.dim));
        await this._sleep(1000);

        let totalDuration = 0;
        for (let i = 0; i < cycles; i++) {
            console.log(c(`\nCycle ${i+1}/${cycles}`, colors.magenta));
            totalDuration += await this._breathingCycle();
        }

        if (this.soundOn) { beep(); beep(); beep(); }
        console.log(c('\n✨ Session complete! ✨', colors.green));
        console.log(`  Cycles: ${cycles}`);
        console.log(`  Duration: ${totalDuration.toFixed(1)} seconds`);
        this.stats.recordSession(cycles, totalDuration);
    }

    showStats() {
        console.log('\n📊 STATISTICS');
        console.log(c('─'.repeat(30), colors.dim));
        console.log(`  Total Sessions: ${this.stats.totalSessions}`);
        console.log(`  Total Cycles:   ${this.stats.totalCycles}`);
        const mins = Math.floor(this.stats.totalTimeSeconds / 60);
        const secs = this.stats.totalTimeSeconds % 60;
        console.log(`  Total Time:     ${mins}m ${secs}s`);
        if (this.stats.sessions.length) {
            const last = this.stats.sessions[this.stats.sessions.length-1];
            console.log(`  Last Session:   ${last.cycles} cycles (${last.durationSeconds}s)`);
            console.log('\n📅 Recent Sessions:');
            this.stats.sessions.slice(-5).forEach(s => {
                const d = s.date.slice(0,16).replace('T', ' ');
                console.log(`  ${d}  ${s.cycles} cycles  ${s.durationSeconds}s`);
            });
        }
    }

    async setTiming() {
        const inhale = await this._askInt(`Inhale duration (default ${this.inhale}): `) || this.inhale;
        const hold = await this._askInt(`Hold duration (default ${this.hold}): `) || this.hold;
        const exhale = await this._askInt(`Exhale duration (default ${this.exhale}): `) || this.exhale;
        this.inhale = inhale;
        this.hold = hold;
        this.exhale = exhale;
        this.stats.setTiming(inhale, hold, exhale);
        console.log(c(`✅ Timing set: ${inhale}s inhale, ${hold}s hold, ${exhale}s exhale`, colors.green));
    }

    toggleSound() {
        this.soundOn = !this.soundOn;
        console.log(c(`🔊 Sound ${this.soundOn ? 'on' : 'off'}`, colors.cyan));
    }

    async resetStats() {
        const ans = await this._ask('⚠️  Reset all statistics? (yes/no): ');
        if (ans.toLowerCase() === 'yes') {
            this.stats.totalSessions = 0;
            this.stats.totalCycles = 0;
            this.stats.totalTimeSeconds = 0;
            this.stats.sessions = [];
            this.stats.save();
            console.log(c('🗑️  Statistics reset.', colors.yellow));
        }
    }

    // ─── Helpers ──────────────────────────────────────────────────────────

    _ask(prompt) {
        return new Promise(resolve => this.rl.question(prompt, resolve));
    }

    async _askInt(prompt) {
        const ans = await this._ask(prompt);
        const num = parseInt(ans.trim());
        return isNaN(num) ? null : num;
    }

    async _showMenu() {
        console.log('\n' + c('═'.repeat(50), colors.cyan));
        console.log(c('🌬️ BREATH TRAINER', colors.bright + colors.cyan));
        console.log(c('═'.repeat(50), colors.cyan));
        console.log(`  4-7-8: Inhale ${this.inhale}s, Hold ${this.hold}s, Exhale ${this.exhale}s`);
        console.log(`  Sound: ${this.soundOn ? '🔊 On' : '🔇 Off'}`);
        console.log(c('═'.repeat(50), colors.cyan));
        console.log('  1. 🌬️ Start Session (4 cycles)');
        console.log('  2. 🌬️ Custom Session (choose cycles)');
        console.log('  3. 📊 Statistics');
        console.log('  4. ⏱️ Set Timings');
        console.log('  5. 🔇 Toggle Sound');
        console.log('  6. 🗑️ Reset Statistics');
        console.log('  0. 🚪 Exit');
        console.log(c('═'.repeat(50), colors.cyan));
    }

    async run() {
        console.clear();
        console.log(c('\n🌬️ Breath Trainer – 4-7-8 Relaxation', colors.bright + colors.cyan));
        console.log(c('Breathe in peace, hold calm, exhale stress.', colors.dim));

        while (true) {
            await this._showMenu();
            const choice = await this._ask('Your choice: ');
            switch (choice.trim()) {
                case '1':
                    await this.runSession(4);
                    break;
                case '2': {
                    const cycles = await this._askInt('Number of cycles (default 4): ') || 4;
                    await this.runSession(Math.max(1, cycles));
                    break;
                }
                case '3':
                    this.showStats();
                    break;
                case '4':
                    await this.setTiming();
                    break;
                case '5':
                    this.toggleSound();
                    break;
                case '6':
                    await this.resetStats();
                    break;
                case '0':
                    console.log(c('👋 Breathe deeply! Goodbye!', colors.cyan));
                    this.rl.close();
                    return;
                default:
                    console.log(c('❌ Invalid choice.', colors.red));
            }
            if (choice !== '0') {
                console.log('\nPress Enter to continue...');
                await this._ask('');
            }
        }
    }
}

// ─── Main ────────────────────────────────────────────────────────────────────

const main = async () => {
    try {
        const app = new BreathTrainer();
        await app.run();
    } catch (e) {
        console.error(c(`❌ Unexpected error: ${e.message}`, colors.red));
        process.exit(1);
    }
};

main();

File: breath_trainer.py
# breath_trainer.py
#!/usr/bin/env python3
"""
🌬️ Breath Trainer – 4-7-8 Relaxation Technique (Python Edition)
Features: guided breathing, visual progress bars, session stats, custom timing, sound alerts
"""

import json
import os
import sys
import time
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional

try:
    from rich.console import Console
    from rich.panel import Panel
    from rich.prompt import Prompt, IntPrompt, Confirm
    from rich.progress import Progress, BarColumn, TextColumn
    from rich import box
    RICH_AVAILABLE = True
except ImportError:
    RICH_AVAILABLE = False
    print("⚠️  Install 'rich' for enhanced UI: pip install rich")


# ─── Colors ──────────────────────────────────────────────────────────────────

def c(text: str, color: str) -> str:
    colors = {
        "reset": "\033[0m", "bright": "\033[1m", "dim": "\033[2m",
        "red": "\033[31m", "green": "\033[32m", "yellow": "\033[33m",
        "blue": "\033[34m", "magenta": "\033[35m", "cyan": "\033[36m"
    }
    return f"{colors.get(color, '')}{text}{colors['reset']}"


# ─── Sound ──────────────────────────────────────────────────────────────────

def beep(frequency: int = 800, duration: int = 200):
    """Cross‑platform beep."""
    try:
        if sys.platform == "win32":
            import winsound
            winsound.Beep(frequency, duration)
        else:
            # Use terminal bell as fallback
            print('\a', end='', flush=True)
    except Exception:
        pass


# ─── Data Manager ──────────────────────────────────────────────────────────

class StatsManager:
    DATA_DIR = Path.home() / ".breath_trainer"
    DATA_FILE = DATA_DIR / "stats.json"

    def __init__(self):
        self.total_sessions = 0
        self.total_cycles = 0
        self.total_time_seconds = 0
        self.sessions: List[Dict] = []
        self.custom_timing = {"inhale": 4, "hold": 7, "exhale": 8}
        self._load()

    def _load(self):
        if self.DATA_FILE.exists():
            try:
                with open(self.DATA_FILE, 'r') as f:
                    data = json.load(f)
                    self.total_sessions = data.get("total_sessions", 0)
                    self.total_cycles = data.get("total_cycles", 0)
                    self.total_time_seconds = data.get("total_time_seconds", 0)
                    self.sessions = data.get("sessions", [])
                    self.custom_timing = data.get("custom_timing", {"inhale": 4, "hold": 7, "exhale": 8})
            except Exception:
                pass

    def save(self):
        self.DATA_DIR.mkdir(parents=True, exist_ok=True)
        data = {
            "total_sessions": self.total_sessions,
            "total_cycles": self.total_cycles,
            "total_time_seconds": self.total_time_seconds,
            "sessions": self.sessions,
            "custom_timing": self.custom_timing,
        }
        with open(self.DATA_FILE, 'w') as f:
            json.dump(data, f, indent=2)

    def record_session(self, cycles: int, duration: float):
        self.total_sessions += 1
        self.total_cycles += cycles
        self.total_time_seconds += int(duration)
        self.sessions.append({
            "date": datetime.now().isoformat(),
            "cycles": cycles,
            "duration_seconds": int(duration),
        })
        if len(self.sessions) > 50:
            self.sessions = self.sessions[-50:]
        self.save()

    def set_timing(self, inhale: int, hold: int, exhale: int):
        self.custom_timing = {"inhale": inhale, "hold": hold, "exhale": exhale}
        self.save()


# ─── Breath Trainer ──────────────────────────────────────────────────────

class BreathTrainer:
    def __init__(self):
        self.console = Console() if RICH_AVAILABLE else None
        self.stats = StatsManager()
        self.inhale = self.stats.custom_timing["inhale"]
        self.hold = self.stats.custom_timing["hold"]
        self.exhale = self.stats.custom_timing["exhale"]
        self.sound_on = True

    def _show_stage(self, stage: str, duration: float, color: str, emoji: str):
        """Display a breathing stage with a progress bar."""
        if self.console:
            with Progress(
                TextColumn(f"[bold {color}]{emoji} {{task.description}}", justify="center"),
                BarColumn(bar_width=40, style=color, complete_style=color),
                TextColumn("[progress.percentage]{task.percentage:>3.0f}%", style=color),
                transient=True,
            ) as progress:
                task = progress.add_task(f"{stage}  ", total=duration * 10)
                for _ in range(int(duration * 10)):
                    time.sleep(0.1)
                    progress.update(task, advance=1)
        else:
            # Simple ASCII bar without rich
            total_steps = int(duration * 2)
            for i in range(total_steps + 1):
                pct = i / total_steps * 100
                filled = int(i / total_steps * 30)
                bar = "█" * filled + "░" * (30 - filled)
                sys.stdout.write(f"\r  {emoji} {stage} {bar} {pct:3.0f}%")
                sys.stdout.flush()
                time.sleep(0.5)
            print()

    def _breathing_cycle(self) -> float:
        """Run one 4-7-8 cycle and return the duration."""
        start = time.time()
        # Inhale
        beep(600, 100) if self.sound_on else None
        self._show_stage("Breathe In...", self.inhale, "blue", "🌬️")
        # Hold
        beep(800, 100) if self.sound_on else None
        self._show_stage("Hold...", self.hold, "yellow", "⏸️")
        # Exhale
        beep(400, 150) if self.sound_on else None
        self._show_stage("Exhale...", self.exhale, "green", "🌊")
        return time.time() - start

    def run_session(self, cycles: int = 4):
        """Run a full breathing session."""
        if self.console:
            self.console.print(Panel.fit(f"[bold cyan]🌬️ 4-7-8 Breathing Session[/bold cyan]\n{cycles} cycles", border_style="cyan"))
        else:
            print(c(f"\n🌬️ 4-7-8 Breathing Session – {cycles} cycles", "bright"))

        beep(1000, 200) if self.sound_on else None
        print(c("\nGet ready...", "dim"))
        time.sleep(1)

        total_duration = 0
        for i in range(cycles):
            if self.console:
                self.console.print(f"[bold magenta]Cycle {i+1}/{cycles}[/bold magenta]")
            else:
                print(c(f"\nCycle {i+1}/{cycles}", "magenta"))
            total_duration += self._breathing_cycle()

        beep(1200, 300) if self.sound_on else None
        if self.console:
            self.console.print("[bold green]✨ Session complete! ✨[/bold green]")
            self.console.print(f"  Cycles: {cycles}")
            self.console.print(f"  Duration: {total_duration:.1f} seconds")
        else:
            print(c("\n✨ Session complete! ✨", "green"))
            print(f"  Cycles: {cycles}")
            print(f"  Duration: {total_duration:.1f} seconds")

        self.stats.record_session(cycles, total_duration)

    def show_stats(self):
        """Display session statistics."""
        if self.console:
            from rich.table import Table
            table = Table(title="📊 Breathing Statistics", box=box.ROUNDED)
            table.add_column("Metric", style="cyan")
            table.add_column("Value", style="green")
            table.add_row("Total Sessions", str(self.stats.total_sessions))
            table.add_row("Total Cycles", str(self.stats.total_cycles))
            table.add_row("Total Time", f"{self.stats.total_time_seconds // 60}m {self.stats.total_time_seconds % 60}s")
            if self.stats.sessions:
                last = self.stats.sessions[-1]
                table.add_row("Last Session", f"{last['cycles']} cycles ({last['duration_seconds']}s)")
            self.console.print(table)

            if self.stats.sessions:
                hist_table = Table(title="📅 Recent Sessions", box=box.MINIMAL)
                hist_table.add_column("Date", style="dim")
                hist_table.add_column("Cycles", justify="right")
                hist_table.add_column("Duration", justify="right")
                for s in self.stats.sessions[-5:]:
                    date = s["date"][:16].replace("T", " ")
                    hist_table.add_row(date, str(s["cycles"]), f"{s['duration_seconds']}s")
                self.console.print(hist_table)
        else:
            print("\n📊 STATISTICS")
            print(c("─"*30, "dim"))
            print(f"  Total Sessions: {self.stats.total_sessions}")
            print(f"  Total Cycles:   {self.stats.total_cycles}")
            print(f"  Total Time:     {self.stats.total_time_seconds // 60}m {self.stats.total_time_seconds % 60}s")
            if self.stats.sessions:
                last = self.stats.sessions[-1]
                print(f"  Last Session:   {last['cycles']} cycles ({last['duration_seconds']}s)")
                print("\n📅 Recent Sessions:")
                for s in self.stats.sessions[-5:]:
                    date = s["date"][:16].replace("T", " ")
                    print(f"  {date}  {s['cycles']} cycles  {s['duration_seconds']}s")

    def set_timing(self):
        if self.console:
            inhale = IntPrompt.ask("Inhale duration (seconds)", default=self.inhale)
            hold = IntPrompt.ask("Hold duration (seconds)", default=self.hold)
            exhale = IntPrompt.ask("Exhale duration (seconds)", default=self.exhale)
        else:
            try:
                inhale = int(input(f"Inhale duration (default {self.inhale}): ") or self.inhale)
                hold = int(input(f"Hold duration (default {self.hold}): ") or self.hold)
                exhale = int(input(f"Exhale duration (default {self.exhale}): ") or self.exhale)
            except ValueError:
                print(c("Invalid input. Keeping current values.", "yellow"))
                return
        self.inhale = inhale
        self.hold = hold
        self.exhale = exhale
        self.stats.set_timing(inhale, hold, exhale)
        print(c(f"✅ Timing set: {inhale}s inhale, {hold}s hold, {exhale}s exhale", "green"))

    def toggle_sound(self):
        self.sound_on = not self.sound_on
        state = "on" if self.sound_on else "off"
        print(c(f"🔊 Sound {state}", "cyan"))

    def reset_stats(self):
        if self.console:
            if not Confirm.ask("⚠️  Reset all statistics?"):
                return
        else:
            if input("⚠️  Reset all statistics? (yes/no): ").strip().lower() != "yes":
                return
        self.stats.total_sessions = 0
        self.stats.total_cycles = 0
        self.stats.total_time_seconds = 0
        self.stats.sessions = []
        self.stats.save()
        print(c("🗑️  Statistics reset.", "yellow"))

    # ─── Menu ──────────────────────────────────────────────────────────────

    def _show_menu(self):
        if self.console:
            menu = f"""
[bold cyan]🌬️ Breath Trainer[/bold cyan]
  4-7-8 Technique: Inhale {self.inhale}s, Hold {self.hold}s, Exhale {self.exhale}s
  Sound: {"🔊 On" if self.sound_on else "🔇 Off"}

  [1] 🌬️ Start Session (4 cycles)
  [2] 🌬️ Custom Session (choose cycles)
  [3] 📊 Statistics
  [4] ⏱️ Set Timings
  [5] 🔇 Toggle Sound
  [6] 🗑️ Reset Statistics
  [0] 🚪 Exit
"""
            self.console.print(Panel(menu, border_style="blue"))
        else:
            print("\n" + "-"*50)
            print(f"🌬️ 4-7-8: Inhale {self.inhale}s, Hold {self.hold}s, Exhale {self.exhale}s")
            print(f"   Sound: {'🔊 On' if self.sound_on else '🔇 Off'}")
            print("-"*50)
            print("  1. 🌬️ Start Session (4 cycles)")
            print("  2. 🌬️ Custom Session (choose cycles)")
            print("  3. 📊 Statistics")
            print("  4. ⏱️ Set Timings")
            print("  5. 🔇 Toggle Sound")
            print("  6. 🗑️ Reset Statistics")
            print("  0. 🚪 Exit")
            print("-"*50)

    def _get_choice(self) -> str:
        if self.console:
            return Prompt.ask("Your choice", choices=["0","1","2","3","4","5","6"])
        return input("Your choice: ").strip()

    def run(self):
        if self.console:
            self.console.print(Panel.fit("[bold cyan]🌬️ Breath Trainer – 4-7-8 Relaxation[/bold cyan]", border_style="cyan"))
        else:
            print(c("\n🌬️ Breath Trainer – 4-7-8 Relaxation", "bright"))
            print(c("Breathe in peace, hold calm, exhale stress.", "dim"))

        while True:
            self._show_menu()
            choice = self._get_choice()
            if choice == "1":
                self.run_session(4)
            elif choice == "2":
                if self.console:
                    cycles = IntPrompt.ask("Number of cycles", default=4)
                else:
                    try:
                        cycles = int(input("Number of cycles (default 4): ") or 4)
                    except ValueError:
                        cycles = 4
                self.run_session(max(1, cycles))
            elif choice == "3":
                self.show_stats()
            elif choice == "4":
                self.set_timing()
            elif choice == "5":
                self.toggle_sound()
            elif choice == "6":
                self.reset_stats()
            elif choice == "0":
                print(c("👋 Breathe deeply! Goodbye!", "cyan"))
                break
            else:
                print(c("❌ Invalid choice.", "red"))

            if choice != "0":
                if self.console:
                    self.console.print("\n[dim]Press Enter to continue...[/dim]")
                    input()
                else:
                    input("\nPress Enter to continue...")


def main():
    try:
        app = BreathTrainer()
        app.run()
    except KeyboardInterrupt:
        print("\n👋 Breathe deeply!")
        sys.exit(0)
    except Exception as e:
        print(c(f"❌ Unexpected error: {e}", "red"))
        sys.exit(1)

if __name__ == "__main__":
    main()
